### ISSUE TITLE
[Library Stabilization] `bigint.mod` int implementation

### ISSUE THREAD

**Issue by lydia-duncan** [chapel-lang/chapel#17713]
The documentation for the bigint version in GMP says that the sign of the bigint is ignored, but for our int implementation we cast any negative values to a uint to call the version that returns a remainder as well.  In other cases we convert ints to bigints and call the bigint version, should we do the same for ints here instead?  That might involve splitting into separate functions since that version wouldn't return a remainder.  Or should the uint version do the same thing instead of calling mpz_mod_ui (which is the uint specific implementation)?

**Comment by lydia-duncan** [comment]
Michael says:
> I'm not sure, but I expect there is probably surprising behavior here when the argument changes between int/uint/bigint.

**Comment by bradcray** [comment]
As I understand it, the sign of the second argument (`m` in `n mod m`) is ignored in GMP, not the first?  I know that our implementation is careful about negative `n` values since C's `%` is fairly non-mathematical in that case, but I don't have any intuition as to what a negative `m` value should mean for a modulus operator, mathematically (specifically, I think of mod m as meaning return a value from `0..m-1`, but for negative `m`...?).  Doing a quick google search hasn't helped me yet...

**Comment by lydia-duncan** [comment]
> As I understand it, the sign of the second argument (`m` in `n mod m`) is ignored in GMP, not the first? I know that our implementation is careful about negative `n` values since C's `%` is fairly non-mathematical in that case,

Right.  The two versions of `bigint.mod` always take a `bigint` for the `n` portion of the computation, this is only referring to the `m` portion. 

> but I don't have any intuition as to what a negative `m` value should mean for a modulus operator, mathematically (specifically, I think of mod m as meaning return a value from `0..m-1`, but for negative `m`...?). Doing a quick google search hasn't helped me yet...

Here are a few computations and their results according to WolframAlpha:
- 5 mod -2 = -1
- 5 mod -3 = -1
- 5 mod -4 = -3
- 5 mod -5 = 0
- 5 mod -6 = -1
- 7 mod -3 = -2
- 7 mod -4 = -1
- 7 mod -5 = -3

Now, it sounds like what GMP does would not get those results and potentially like the results it gets would be similar to what we'd get by the conversion we perform.  That said, it seems potentially more consistent to rely on the adjustment GMP performs than to guess what it does for its type and try to do that ourselves for ints.

**Comment by lydia-duncan** [comment]
After the module re-review meeting, I had a task to take a look at how the various mods differ in our implementation.  Today, `%`, `BigInteger.%` and `bigint.mod` (both the `(bigint, int)` and `(bigint, bigint)` versions) all behave the same, while `Math.mod` is the only one that returns the values I saw from WolframAlpha.

<details>
Code:
```chapel
use BigInteger;
use Math;

// This test is to check and compare the behavior of mod when dealing with
// negative numbers.
var a1 = 5;
var a2 = 7;
var a3: bigint = a1;
var a4: bigint = a2;

var b1 = -2;
var b2 = -3;
var b3 = -4;
var b4 = -5;
var b5 = -6;

// Check %
writeln("% cases");

var percentRes1 = a1 % b1; // 5 mod -2
writeln(percentRes1); // 1
var percentRes2 = a1 % b2; // 5 mod -3
writeln(percentRes2); // 2
var percentRes3 = a1 % b3; // 5 mod -4
writeln(percentRes3); // 1
var percentRes4 = a1 % b4; // 5 mod -5
writeln(percentRes4); // 0
var percentRes5 = a1 % b5; // 5 mod -6
writeln(percentRes5); // 5
var percentRes6 = a2 % b2; // 7 mod -3
writeln(percentRes6); // 1
var percentRes7 = a2 % b3; // 7 mod -4
writeln(percentRes7); // 3
var percentRes8 = a2 % b4; // 7 mod -5
writeln(percentRes8); // 2


// Check mod
writeln("mod() cases");

var modRes1 = mod(a1, b1); // 5 mod -2
writeln(modRes1); // -1
var modRes2 = mod(a1, b2); // 5 mod -3
writeln(modRes2); // -1
var modRes3 = mod(a1, b3); // 5 mod -4
writeln(modRes3); // -3
var modRes4 = mod(a1, b4); // 5 mod -5
writeln(modRes4); // 0
var modRes5 = mod(a1, b5); // 5 mod -6
writeln(modRes5); // -1
var modRes6 = mod(a2, b2); // 7 mod -3
writeln(modRes6); // -2
var modRes7 = mod(a2, b3); // 7 mod -4
writeln(modRes7); // -1
var modRes8 = mod(a2, b4); // 7 mod -5
writeln(modRes8); // -3


// Check bigint.%
writeln("bigint.% cases");

var bigPercentRes1 = a3 % b1; // 5 mod -2
writeln(bigPercentRes1); // 1
var bigPercentRes2 = a3 % b2; // 5 mod -3
writeln(bigPercentRes2); // 2
var bigPercentRes3 = a3 % b3; // 5 mod -4
writeln(bigPercentRes3); // 1
var bigPercentRes4 = a3 % b4; // 5 mod -5
writeln(bigPercentRes4); // 0
var bigPercentRes5 = a3 % b5; // 5 mod -6
writeln(bigPercentRes5); // 5
var bigPercentRes6 = a4 % b2; // 7 mod -3
writeln(bigPercentRes6); // 1
var bigPercentRes7 = a4 % b3; // 7 mod -4
writeln(bigPercentRes7); // 3
var bigPercentRes8 = a4 % b4; // 7 mod -5
writeln(bigPercentRes8); // 2


// Check bigint.mod
writeln("bigint.mod() cases");

var bigModRes: bigint;
bigModRes.mod(a3, b1); // 5 mod -2
writeln(bigModRes); // ?
bigModRes.mod(a3, b2); // 5 mod -3
writeln(bigModRes); // ?
bigModRes.mod(a3, b3); // 5 mod -4
writeln(bigModRes); // ?
bigModRes.mod(a3, b4); // 5 mod -5
writeln(bigModRes); // 0
bigModRes.mod(a3, b5); // 5 mod -6
writeln(bigModRes); // ?
bigModRes.mod(a4, b2); // 7 mod -3
writeln(bigModRes); // ?
bigModRes.mod(a4, b3); // 7 mod -4
writeln(bigModRes); // ?
bigModRes.mod(a4, b4); // 7 mod -5
writeln(bigModRes); // ?
```

Results:
```
% cases
1
2
1
0
5
1
3
2
mod() cases
-1
-1
-3
0
-1
-2
-1
-3
bigint.% cases
1
2
1
0
5
1
3
2
bigint.mod() cases
1
2
1
0
5
1
3
2
```
</details>

I think that

**Comment by bradcray** [comment]
On the basis of that investigation, I'd be inclined to make `bigint.mod` match `Math.mod` and keep `%` on `bigint` as it is.

It looks like you were typing a proposal but got cut off:

> ```
> I think that 
> ```

**Comment by lydia-duncan** [comment]
> It looks like you were typing a proposal but got cut off:

Thanks, I've removed it from there (got distracted by the large block of code)

This will likely mean updating both the `mod(bigint, bigint)` and `mod(bigint, int)` versions.

**Comment by jeremiah-corrado** [comment]
Looking at the current implementation a bit closer, I am wondering if the interface of the bigint.mod methods should be changed slightly in addition to updating their behavior to match `Math.mod()`. 

Currently, `bigint.mod(bigint, integral)` returns the result in a `uint`, while `bigint.mod(bigint, bigint)` does not return anything. It looks like the interface is exposing the analogous [GMP methods](https://gmplib.org/manual/Integer-Division) in the following way:
- `void mpz_mod (mpz_t r, const mpz_t n, const mpz_t d)` -> `bigint.mod(a: bigint, b: bigint)`
- `unsigned long int mpz_mod_ui (mpz_t r, const mpz_t n, unsigned long int d)` -> `bigint.mod(a: bigint, b: integral): uint`

So it seems like the `uint` return value is present to mirror the behavior of the `_ui` variant of the mod function.

However, I am wondering if it would be a good idea to remove the return value from the integral overlaod. This is for a few reasons:
- `Math.mod()` behaves uniformly for any combination of input types
- The bigint interface doesn't currently expose any of GMP's other `_ui` methods in this way (it doesn't expose the `div_*_ui` methods at all)
- In order to mirror `Math.mod()`'s behavior we probably can't use the GMP mod methods anyway because they ignore negative arguments, so `a % b = a - b * floor(a / b)` will likely have to be implemented using `mpz_fdiv_q` instead. (The % operator on bigint will still use `mpz_mod`)

**Comment by lydia-duncan** [comment]
> Currently, `bigint.mod(bigint, integral)` returns the result in a `uint`, while `bigint.mod(bigint, bigint)` does not return anything. It looks like the interface is exposing the analogous [GMP methods](https://gmplib.org/manual/Integer-Division) in the following way:
> 
>     * `void mpz_mod (mpz_t r, const mpz_t n, const mpz_t d)` -> `bigint.mod(a: bigint, b: bigint)`
> 
>     * `unsigned long int mpz_mod_ui (mpz_t r, const mpz_t n, unsigned long int d)` -> `bigint.mod(a: bigint, b: integral): uint`
> 
> 
> So it seems like the `uint` return value is present to mirror the behavior of the `_ui` variant of the mod function.

My interpretation was that that reason the `bigint.mod(bigint, bigint)` version doesn't return is because it is putting the result in `this`.  We do this because creating a new `bigint` is expensive, memory-wise (see #17699 for the discussion on that).

I think it's okay for the interface to differ between uints and bigints in this regard because of this, but it would be good for others to weigh in.

> * The bigint interface doesn't currently expose any of GMP's other `_ui` methods in this way (it doesn't expose the `div_*_ui` methods at all)

See #17712 (which isn't for 2.0 because it's purely new functionality)

> * In order to mirror `Math.mod()`'s behavior we probably can't use the GMP mod methods anyway because they ignore negative arguments, so `a % b = a - b * floor(a / b)` will likely have to be implemented using `mpz_fdiv_q` instead. (The % operator on bigint will still use `mpz_mod`)

That seems reasonable to me

**Comment by jeremiah-corrado** [comment]
> That'd help but only if mpz_fdiv_ui didn't take in a bigint for the result. If it does, we'd just have to create a local bigint anyways to call it and that's strictly worse than taking one the user's already provided

It does not, this is the header from GMP's docs: `unsigned long int mpz_fdiv_ui (const mpz_t n, unsigned long int d)`. So I think we could make a more memory efficient `bigint.mod(bigint, integral): int` that does not update `this`. My only hesitation, is that it could be confusing for users since none of the other non-operator style methods in BigInteger behave this way.

On a separate note, I looked further into the behavior of the various mod methods, and found that `bigint.%()` (which is implemented directly with `mpz_mod`) does not match the behavior of the % operator for integers. It also doesn't match the behavior of either mathematically correct version of the mod operator. Here is a direct comparison of our mod methods for the example of `(+/-)9 % (+/-)4`:

| A | B | A % B (int) | Math.mod(A, B) | A % B (bigint) | bigint.mod(A, B)** |
|--|--|--|--|--|--|
| _ | _ | Trunc-mod* | Floor-mod | Neither | Neither |
|9|4|1|1|1|1|
|9|-4|1|-3|1|1|
|-9|4|-1|3|3|3|
|-9|-4|-1|-1|3|3|

*C implements % as a truncating mod
**bigint's mod operator and method are currently the same. The initial goal here was to update bigint.mod()'s behavior to match math.mod()

[This page](https://www.omnicalculator.com/math/modulo-of-negative-numbers#how-does-modulo-work-with-negative-numbers) does a good job of explaining the difference between a **truncating-modulo** and a **flooring-modulo**.

I propose that we also update `bigint.%()` and `bigint.%=()` to behave like a truncating modulo. While `bigint.mod()` should still be updated to behave like a flooring mod.

**Comment by mppf** [comment]
From the linked page, I think the way `bigint.mod` is behaving is the number theoretic way (where truncating-modulo and flooring-module are common alternatives in programming languages). However, I think it's reasonable for us to make `bigint.mod` consistent with `Math.mod` and to make `bigint.%` consistent with `%` on integers.

**Comment by jeremiah-corrado** [comment]
> From the linked page, I think the way bigint.mod is behaving is the number theoretic way (where truncating-modulo and flooring-module are common alternatives in programming languages).

I did not realize that, thanks for pointing that out! 

If we do go ahead and unify the behavior between the `integer` and `bigint` mods, should we also somehow expose the `mpz_mod` function in the api given that it has some utility to number theory applications? This could be another method called `nt_mod` or something similar.

**Comment by mppf** [comment]
> should we also somehow expose the mpz_mod function in the api given that it has some utility to number theory applications? 

Probably best to leave that for the future (and wait for somebody to request it).

**Comment by lydia-duncan** [comment]
`mpz_mod` will still be visible from the `GMP` module

**Comment by jeremiah-corrado** [comment]
> Probably best to leave that for the future (and wait for somebody to request it).

Just to make a quick note of how this might be done in the future:

`bigint.mod()` could be modified to take an optional argument that selects between the three modulo variants: flooring, truncating, and "number theoretic". The default would be flooring S.T. this addition would not be a breaking change, and the default behavior would still match with `Math.mod()`. 

This could be done either using BigInteger's existing `round` enum -- with the benefit that `mod`'s new interface would match nicely with the other division functions that expose the rounding options in the same way. Or a new enum could be created to represent the three modulo variants -- with the benefit that the names for each behavior would be more self-documenting.

**Comment by lydia-duncan** [comment]
I think this was resolved by #20019 - @jeremiah-corrado, do you agree?

**Comment by jeremiah-corrado** [comment]
Yeah it was. I just forgot to mark this issue as resolved when I merged. Thanks!
